src/lsm/groove.zig

```diff
@@ -707,40 +728,35 @@ pub fn GrooveType(
             lookup: LookupContext = undefined,
 
             fn lookup_start_next(worker: *PrefetchWorker) void {
-                const id = worker.context.id_iterator.next() orelse {
+                const prefetch_key = worker.context.key_iterator.next() orelse {
                     worker.context.worker_finished();
                     return;
                 };
 
-                if (!has_id) {
-                    worker.lookup_with_timestamp(id.*);
-                    return;
-                }
-
-                if (worker.context.groove.ids.lookup_from_memory(
-                    worker.context.snapshot,
-                    id.*,
-                )) |id_tree_value| {
-                    assert(!id_tree_value.tombstone());
-                    lookup_id_callback(&worker.lookup.id, id_tree_value);
-
-                    if (constants.verify) {
-                        // If the id is cached, then we must be prefetching it because the object
-                        // was not also cached.
-                        assert(worker.context.groove.objects.lookup_from_memory(
-                            worker.context.snapshot,
-                            id_tree_value.timestamp,
-                        ) == null);
-                    }
-                } else {
-                    // If not in the LSM tree's cache, the object must be read from disk and added
-                    // to the auxiliary prefetch_objects hash map.
-                    worker.context.groove.ids.lookup_from_levels(
-                        lookup_id_callback,
-                        &worker.lookup.id,
-                        worker.context.snapshot,
-                        id.*,
-                    );
+                // prefetch_enqueue() ensures that the tree's cache is checked before queueing the
+                // object for prefetching. If not in the LSM tree's cache, the object must be read
+                // from disk and added to the auxiliary prefetch_objects hash map.
+                switch (prefetch_key.kind) {
+                    .id => |id| {
+                        if (has_id) {
+                            worker.context.groove.ids.lookup_from_levels_storage(.{
+                                .callback = lookup_id_callback,
+                                .context = &worker.lookup.id,
+                                .snapshot = worker.context.snapshot,
```

Comment:
`worker.lookup.id` is confusing. Find better name.

Reply:
Seems to me that cleaning this up needs a bit more of thinking & possible discussions, I'm thinking about deferring this to a future PR.

cc: @jorangreef

Reply:
Sure, let's defer by opening an issue.

Reply:
Issue for the same: https://github.com/tigerbeetle/tigerbeetle/issues/1058